src/StreamJsonRpc/JsonRpc.cs

```diff
@@ -234,13 +236,83 @@ public static JsonRpc Attach(Stream sendingStream, Stream receivingStream, objec
             Verify.Operation(!this.startedListening, Resources.AttachTargetAfterStartListeningError);
 
-            this.targetRequestMethodToClrMethodMap.Add(Tuple.Create(target, new ReadOnlyDictionary<string, string>(GetRequestMethodToClrMethodMap(target))));
+            var mapping = GetRequestMethodToClrMethodMap(target);
+            lock (this.syncObject)
+            {
+                foreach (var item in mapping)
+                {
+                    if (this.targetRequestMethodToClrMethodMap.TryGetValue(item.Key, out var existingList))
+                    {
+                        // Only add methods that do not have equivalent signatures to what we already have.
+                        foreach (var newMethod in item.Value)
+                        {
+                            if (!existingList.Any(e => e.Signature.Equals(newMethod.Signature)))
+                            {
+                                existingList.Add(newMethod);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        this.targetRequestMethodToClrMethodMap.Add(item.Key, item.Value);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Adds a handler for an RPC method with a given name.
+        /// </summary>
+        /// <param name="rpcMethodName">
+        /// The name of the method as it is identified by the incoming JSON-RPC message.
+        /// It need not match the name of the CLR method/delegate given here.
+        /// </param>
+        /// <param name="handler">
+        /// The method or delegate to invoke when a matching RPC message arrives.
+        /// This method may accept parameters from the incoming JSON-RPC message.
+        /// </param>
+        public void AddLocalRpcMethod(string rpcMethodName, Delegate handler)
+        {
+            this.AddLocalRpcMethod(rpcMethodName, handler.GetMethodInfo(), handler.Target);
+        }
+
+        /// <summary>
+        /// Adds a handler for an RPC method with a given name.
+        /// </summary>
+        /// <param name="rpcMethodName">
+        /// The name of the method as it is identified by the incoming JSON-RPC message.
+        /// It need not match the name of the CLR method/delegate given here.
+        /// </param>
+        /// <param name="handler">
+        /// The method or delegate to invoke when a matching RPC message arrives.
+        /// This method may accept parameters from the incoming JSON-RPC message.
+        /// </param>
+        /// <param name="target">An instance of the type that defines <paramref name="handler"/> which should handle the invocation.</param>
+        public void AddLocalRpcMethod(string rpcMethodName, MethodInfo handler, object target)
+        {
+            Verify.Operation(!this.startedListening, Resources.AttachTargetAfterStartListeningError);
```

Comment:
This will defeat dynamic light up of new methods which may need to occur after we've started listening.

Reply:
True, but that's consistent with behavior before my change since the `AddLocalRpcTarget` method already did this.

IIRC we talked about how we want to enable this scenario, and that's exactly why I added a lock around access to the dictionary we mutate in this method. But actually enabling the scenario deserves a PR of its own, IMO. IIRC it will require a new property or ctor parameter to opt into the feature, since we want to avoid folks unknowingly using `Attach` and then adding target objects/methods and not realizing that they're introducing a race condition for incoming traffic.

Reply:
Opened #62 to track.